lib/Optimizer/Dialect/Quake/QuakeOps.cpp

```diff
@@ -168,6 +189,33 @@ static void printRawIndex(OpAsmPrinter &printer, quake::ExtractRefOp refOp,
 }
 
 namespace {
+// %4 = quake.concat %2, %3 : (!quake.ref, !quake.ref) -> !quake.veq<2>
+// %7 = quake.extract_ref %4[0] : (!quake.veq<2>) -> !quake.ref
+// ───────────────────────────────────────────
+// replace all use with %2
+struct BackwardConcatExtractPattern
+    : public OpRewritePattern<quake::ExtractRefOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::ExtractRefOp extract,
+                                PatternRewriter &rewriter) const override {
+    auto veq = extract.getVeq();
+    auto concatOp = veq.getDefiningOp<quake::ConcatOp>();
+    if (concatOp && extract.hasConstantIndex()) {
+      auto index = extract.getConstantIndex();
+      auto concatQubits = concatOp.getQbits();
+      if (index < concatQubits.size()) {
```

Comment:
This is further complicated by `!quake.veq<?>` arguments, of course. In that case, we shouldn't be doing this, since the k-th ref could come from an indeterminant argument.

Reply:
Will you see if this update fixes the issues? Here I just try to disallow any use of this canonicalizer for a concat op that has any `veq` type arguments.